src/scratchOrgCreate.ts

```diff
@@ -0,0 +1,171 @@
+  orgConfig: Record<string, unknown>;
+  configAggregator: ConfigAggregator;
+  test: boolean;
+  clientSecret: string;
+}
+
+export const scratchOrgCreate = async (options: ScratchOrgCreateOptions): Promise<ScratchOrgCreateResult> => {
+  const logger = await Logger.child('scratchOrgCreate');
+  logger.debug('scratchOrgCreate');
+
+  const {
+    hubOrg,
+    connectedAppConsumerKey,
+    durationDays,
+    nonamespace,
+    noancestors,
+    setdefaultusername,
+    setalias,
+    retry,
+    apiversion,
+    definitionjson,
+    definitionfile,
+    orgConfig,
+    configAggregator,
+    clientSecret,
+  } = options;
+
+  const settingsGenerator = new SettingsGenerator();
+
+  const { scratchOrgInfoPayload, ignoreAncestorIds, warnings } = await getScratchOrgInfoPayload({
+    definitionjson,
+    definitionfile,
+    connectedAppConsumerKey,
+    durationDays,
+    nonamespace,
+    noancestors,
+    orgConfig,
+  });
+
+  const scratchOrgInfo = await generateScratchOrgInfo({
+    hubOrg,
+    scratchOrgInfoPayload,
+    nonamespace,
+    ignoreAncestorIds,
+  });
+
+  // gets the scratch org settings (will use in both signup paths AND to deploy the settings)
+  await settingsGenerator.extract(scratchOrgInfo);
+  logger.debug(`the scratch org def file has settings: ${settingsGenerator.hasSettings()}`);
+
+  // creates the scratch org info in the devhub
+  const scratchOrgInfoRequestResult = await requestScratchOrgCreation(hubOrg, scratchOrgInfo, settingsGenerator);
+
+  const scratchOrgInfoId = ensureString(getString(scratchOrgInfoRequestResult, 'id'));
+  logger.debug(`scratch org has recordId ${scratchOrgInfoId}`);
+
+  const scratchOrgInfoResult = await pollForScratchOrgInfo(options.hubOrg, scratchOrgInfoId, options.wait);
+
+  const signupTargetLoginUrlConfig = await getsSgnupTargetLoginUrlConfig();
+
+  const scratchOrgAuthInfo = await authorizeScratchOrg({
+    scratchOrgInfoComplete: scratchOrgInfoResult,
+    hubOrg,
+    clientSecret,
+    setAsDefault: setdefaultusername,
+    alias: setalias,
+    signupTargetLoginUrlConfig,
+    retry: retry || 0,
+  });
+
+  await Lifecycle.getInstance().emit('scratchOrgCreate', {
+    status: 'getAuthInfo',
+  });
+
+  // we'll need this scratch org connection later;
+  const scratchOrg = await Org.create({ connection: await Connection.create({ authInfo: scratchOrgAuthInfo }) });
+  const username = scratchOrg.getUsername();
+
+  const authInfo = await deploySettingsAndResolveUrl(
+    scratchOrgAuthInfo,
+    apiversion ??
+      (configAggregator.getPropertyValue('apiVersion') as string) ??
+      (await scratchOrg.retrieveMaxApiVersion()),
+    settingsGenerator,
+    scratchOrg
+  );
+
+  logger.trace('Settings deployed to org');
+  /** updating the revision num to zero during org:creation if source members are created during org:create.This only happens for some specific scratch org definition file.*/
+  await updateRevisionCounterToZero(scratchOrg);
+
+  return {
+    username,
```

Comment:
is there anything else a command/consumer might want?  ex: orgId, scratchOrgInfo id/link for the dev hub record, the mdapi output of what got deployed to the org?